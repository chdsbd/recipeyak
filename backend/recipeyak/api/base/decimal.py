from decimal import Decimal
from fractions import Fraction

_MAX_DECIMAL_PLACES = 8


def fmt_decimal(d: Decimal) -> str:
    """
    remove trailing zeros

    Decimal("4.0000") -> "4"
    """
    if d < 1:
        # 0.5 -> 1/2
        # 0.333333 -> 1/3
        num, denom = Fraction(d).limit_denominator().as_integer_ratio()
        return f"{num}/{denom}"
    d = round(d, _MAX_DECIMAL_PLACES)
    if d == d.to_integral():
        return str(d.quantize(Decimal(1)))
    # only allow up to 3 decimal places
    return str(d.normalize().quantize(Decimal("1.000")).normalize())
