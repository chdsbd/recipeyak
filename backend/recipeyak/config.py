import os

AWS_ACCESS_KEY_ID = os.getenv("AWS_ACCESS_KEY_ID", "")
AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY", "")
STORAGE_BUCKET_NAME = os.getenv("STORAGE_BUCKET_NAME", "")
STORAGE_HOSTNAME = os.getenv("STORAGE_HOSTNAME", "")
ABLY_API_KEY = os.getenv("ABLY_API_KEY", "SOME:KEY")
ALGOLIA_APPLICATION_ID = os.getenv("ALGOLIA_APPLICATION_ID", "")
ALGOLIA_ADMIN_API_KEY = os.getenv("ALGOLIA_ADMIN_API_KEY", "")
ALGOLIA_SEARCH_ONLY_API_KEY = os.getenv("ALGOLIA_SEARCH_ONLY_API_KEY", "")

IMAGE_TRANSFORM_FORMAT = os.getenv("IMAGE_TRANSFORM_FORMAT", "twicpics")
