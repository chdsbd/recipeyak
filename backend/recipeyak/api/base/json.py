from collections.abc import ValuesView
from decimal import Decimal
from typing import Any

import orjson
import pydantic

MAX_DECIMAL_PLACES = 8


def fmt_decimal(d: Decimal) -> str:
    """
    remove trailing zeros

    Decimal("4.0000") -> "4"
    """
    d = round(d, MAX_DECIMAL_PLACES)
    if d == d.to_integral():
        return str(d.quantize(Decimal(1)))
    return str(d.normalize())


def default(o: object) -> Any:
    if isinstance(o, Decimal):
        return fmt_decimal(o)
    if isinstance(o, pydantic.BaseModel):
        return o.dict()
    if isinstance(o, ValuesView):
        return list(o)
    raise TypeError


def json_dumps(data: Any, indent: bool = False) -> bytes:
    """
    Serialize `data` to a JSON formatted byte string.
    """
    if data is None:
        return b""
    options = orjson.OPT_NON_STR_KEYS
    if indent:
        options |= orjson.OPT_INDENT_2
    return orjson.dumps(data, default=default, option=options)
