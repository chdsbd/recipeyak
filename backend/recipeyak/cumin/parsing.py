from __future__ import annotations

import decimal
import enum
import itertools
from dataclasses import dataclass
from decimal import Decimal
from enum import Enum

_MALFORMED_UNITS = {"large", "medium", "small", "fresh"}

_quantity_chars = set(" /.0123456789")

_UNICODE_FRACTION_MAPPING = {
    "½": "1/2",
    "⅓": "1/3",
    "⅔": "2/3",
    "¼": "1/4",
    "¾": "3/4",
    "⅕": "1/5",
    "⅖": "2/5",
    "⅗": "3/5",
    "⅘": "4/5",
    "⅙": "1/6",
    "⅚": "5/6",
    "⅐": "1/7",
    "⅛": "1/8",
    "⅜": "3/8",
    "⅝": "5/8",
    "⅞": "7/8",
    "⅑": "1/9",
    "⅒": "1/10",
}


def _unicode_fractions_to_ascii(s: str) -> str:
    """
    convert occurances of unicode fractions like `½` to their ascii
    equivalent `1/2`
    """

    output = ""
    for char in s:
        expanded = _UNICODE_FRACTION_MAPPING.get(char, char)
        # ensure 1¾ becomes 1 3/4, not 13/4
        if output and output[-1].isdigit() and expanded != char:
            output += " "
        output += expanded

    return output


def _max_quantity(quantity: str) -> str:
    """
    take the '4-5' medium button mushrooms and find the max, 5

    1 cup plain whole-milk
    """
    if " to " in quantity:
        return quantity.split(" to ")[-1]
    return quantity.split("-")[-1]


@enum.unique
class Unit(str, Enum):
    # mass/weight
    POUND = "POUND"
    OUNCE = "OUNCE"
    # metric
    GRAM = "GRAM"
    KILOGRAM = "KILOGRAM"

    # volume
    TEASPOON = "TEASPOON"
    TABLESPOON = "TABLESPOON"
    FLUID_OUNCE = "FLUID_OUNCE"
    CUP = "CUP"
    PINT = "PINT"
    QUART = "QUART"
    GALLON = "GALLON"
    # metric volume
    LITER = "LITER"
    MILLILITER = "MILLILITER"

    # covers "pinch", "sprinkle", "dash", "some", etc.
    SOME = "SOME"

    # for instance `1 cucumber`
    NONE = "NONE"

    # "1 bag", we don't parse bag so we just return UNKNOWN and let the caller
    # handle the result
    UNKNOWN = "UNKNOWN"

    def base_unit(self) -> BaseUnit:
        if self in _VOLUME:
            return BaseUnit.VOLUME
        if self in _MASS:
            return BaseUnit.MASS
        if self == Unit.SOME:
            return BaseUnit.SOME
        if self == Unit.NONE:
            return BaseUnit.NONE
        if self == Unit.UNKNOWN:
            return BaseUnit.UNKNOWN
        raise UnhandledCaseError(case=self)

    def __lt__(self, other: Unit) -> bool:  # type: ignore[override]
        if self in _VOLUME:
            return _VOLUME[self] < _VOLUME[other]
        if self in _MASS:
            return _MASS[self] < _MASS[other]
        return False


def _get_unit(val: str) -> Unit:
    val_cased = val.strip()
    val = val_cased.lower()
    if "cup" in val:
        return Unit.CUP
    if val == "kg":
        return Unit.KILOGRAM
    if "gram" in val:
        return Unit.GRAM
    if "gallon" in val:
        return Unit.GALLON
    if "ounce" in val or val == "oz":
        return Unit.OUNCE
    if "milliliter" in val:
        return Unit.MILLILITER
    if val_cased == "t" or "tsp" in val or "teaspoon" in val:
        return Unit.TEASPOON
    if val_cased == "T" or "tbs" in val or "tablespoon" in val:
        return Unit.TABLESPOON
    if "pinch" in val:
        return Unit.SOME
    if "lbs" in val or "lb" in val or "pound" in val:
        return Unit.POUND
    if "quart" in val:
        return Unit.QUART
    if "liter" in val:
        return Unit.LITER
    if val in {"some", "sprinkle", "dash"}:
        return Unit.SOME
    if val == "":
        return Unit.NONE
    return Unit.UNKNOWN


def _fraction_to_decimal(val: str) -> Decimal | None:
    try:
        total = Decimal(0)
        for v in val.strip().split(" "):
            if "/" in v:
                top, bot = v.split("/")
                total += Decimal(top) / Decimal(bot)
            else:
                total += Decimal(v)
        return total
    except decimal.InvalidOperation:
        return None


def _is_unit(val: str) -> bool:
    return _get_unit(val) != Unit.UNKNOWN


@dataclass
class Quantity:
    quantity: Decimal
    unit: Unit
    unknown_unit: str | None = None

    def __add__(self, other: Quantity) -> Quantity:
        if self.unit == other.unit != Unit.UNKNOWN:
            return Quantity(quantity=self.quantity + other.quantity, unit=self.unit)
        if (
            self.unit == other.unit == Unit.UNKNOWN
            and self.unknown_unit == other.unknown_unit
        ):
            return Quantity(
                quantity=self.quantity + other.quantity,
                unit=Unit.UNKNOWN,
                unknown_unit=self.unknown_unit,
            )
        if self.unit == Unit.SOME:
            return other
        if other.unit == Unit.SOME:
            return self
        if self.unit.base_unit() == other.unit.base_unit():
            if self.unit == Unit.UNKNOWN:
                raise IncompatibleUnitError(units=(self.unit, other.unit))
            unit_lookup = _VOLUME if self.unit in _VOLUME else _MASS
            smallest_unit = self.unit if self.unit < other.unit else other.unit
            quantity = self.to_base_unit().quantity + other.to_base_unit().quantity
            return Quantity(
                quantity=quantity / unit_lookup[smallest_unit], unit=smallest_unit
            )
        raise IncompatibleUnitError(units=(self.unit, other.unit))

    def __str__(self) -> str:
        if self.unit == Unit.SOME:
            return "some"
        if self.unit == Unit.NONE:
            return str(self.quantity)
        unit = str(self.unit)
        return f"{self.quantity} {unit}"

    def to_base_unit(self) -> Quantity:
        if self.unit.base_unit() == BaseUnit.VOLUME:
            return Quantity(
                unit=Unit.MILLILITER, quantity=_VOLUME[self.unit] * self.quantity
            )
        if self.unit.base_unit() == BaseUnit.MASS:
            return Quantity(unit=Unit.GRAM, quantity=_MASS[self.unit] * self.quantity)
        return self


def _parse_quantity(val: str) -> Quantity:
    value = iter(_unicode_fractions_to_ascii(_max_quantity(val).strip()))

    quantity = ""
    unit_str = ""
    for c in value:
        if c in _quantity_chars:
            quantity += c
        else:
            unit_str += c
            break

    for c in value:
        if c in _quantity_chars:
            break
        # handle case of two units and quantities next to each other
        unit_str += c

    # strip out misplaced words, e.g., `1 large` `lemon` instead of `1` `large lemon`
    if unit_str in _MALFORMED_UNITS:
        unit_str = ""

    unit = _get_unit(unit_str)

    unknown_unit = None
    if unit == Unit.UNKNOWN:
        unknown_unit = unit_str

    return Quantity(
        quantity=_fraction_to_decimal(quantity) or Decimal(1),
        unit=unit,
        unknown_unit=unknown_unit,
    )


@dataclass
class UnhandledCaseError(Exception):
    case: Unit


@dataclass
class IncompatibleUnitError(Exception):
    units: tuple[Unit, Unit]


@enum.unique
class BaseUnit(Enum):
    MASS = "MASS"
    VOLUME = "VOLUME"
    SOME = "SOME"
    NONE = "NONE"
    UNKNOWN = "UNKNOWN"


_TEASPOON_ML = Decimal(4.92892)
_TABLESPOON_ML = 3 * _TEASPOON_ML
_FLUID_OUNCE_ML = 2 * _TABLESPOON_ML
_CUP_ML = 8 * _FLUID_OUNCE_ML
_PINT_ML = 2 * _CUP_ML
_QUART_ML = 2 * _PINT_ML
_GALLON_ML = 4 * _QUART_ML


_OUNCE_GRAM = Decimal(28.34952)

_MASS: dict[Unit, Decimal] = {
    Unit.GRAM: Decimal(1),
    Unit.OUNCE: _OUNCE_GRAM,
    Unit.POUND: 16 * _OUNCE_GRAM,
    Unit.KILOGRAM: Decimal(1000),
}


_VOLUME: dict[Unit, Decimal] = {
    Unit.MILLILITER: Decimal(1),
    Unit.TEASPOON: _TEASPOON_ML,
    Unit.TABLESPOON: _TABLESPOON_ML,
    Unit.FLUID_OUNCE: _FLUID_OUNCE_ML,
    Unit.CUP: _CUP_ML,
    Unit.PINT: _PINT_ML,
    Unit.QUART: _QUART_ML,
    Unit.LITER: Decimal(1000),
    Unit.GALLON: _GALLON_ML,
}


def parse_quantity(val: str) -> Quantity:
    """
    handle "3 Tablespoon + 1 teaspoon" format
    """
    quantities = val.split("+")
    a = _parse_quantity(quantities[0])
    for q in quantities[1:]:
        a += _parse_quantity(q)
    return a


@dataclass(frozen=True, slots=True)
class _IngredientResult:
    quantity: str
    name: str
    description: str = ""
    optional: bool = False


# TODO(sbdchd): these should be all the units available!!!
_UNIT_TO_ALIASES: dict[Unit, list[str]] = {
    Unit.CUP: ["cup", "cups"],
    Unit.KILOGRAM: ["kg", "kgs", "kilogram", "kilograms"],
    Unit.GRAM: ["gram", "grams", "g"],
    Unit.GALLON: ["gal", "gallon", "gallons"],
    Unit.OUNCE: ["oz", "ounce", "ounces"],
    Unit.MILLILITER: ["millilter", "ml"],
    Unit.TEASPOON: ["tsp", "teaspoon", "teaspoons", "t"],
    Unit.TABLESPOON: ["tablespoon", "tablespoons", "tbsp", "tbs", "T"],
    Unit.POUND: [
        "pound",
        "pounds",
        "lbs",
        "lb",
    ],
    Unit.SOME: [
        "pinch",
        "pinch of",
        "some",
        "sprinkle",
        "dash",
    ],
    Unit.QUART: [
        "quart",
        "quarts",
    ],
    Unit.LITER: [
        "liter",
        "liters",
        "l",
    ],
}

# NOTE(sbdchd): we add some whitespace so when we search the ingredient for the quantity
_units_ws = tuple(
    x + " " for x in itertools.chain.from_iterable(_UNIT_TO_ALIASES.values())
)

_larger_to_smaller_units_ws = sorted(_units_ws, key=lambda x: -len(x))


def _starts_with(string: str, prefixes: tuple[str, ...] | str) -> bool:
    """
    case insensitive str.starts_with
    """
    if isinstance(prefixes, str):
        prefixes = (prefixes,)
    for prefix in prefixes:
        if string[: len(prefix)].casefold() == prefix.casefold():
            return True
    return False


def _parse_quantity_name(text: str) -> tuple[str, str]:
    """
    examples:
      1 cup flour -> 1 cup, flour
      fine sea salt -> some, fine sea salt
      1 tablespoon or 1/3 teaspoon foo bar buzz
      1 1/2 teaspoons minced fresh thyme (or 1/2 teaspoon dried thyme)
      Chopped fresh parsley, for serving (optional)
    """

    value = _unicode_fractions_to_ascii(text.strip())
    quantity = ""
    name = ""

    idx = 0
    in_quantity = True

    while idx < len(value):
        c = value[idx]
        # parse digit like characters into a quantity or name
        if c.isdigit() or c in {"/", ".", "-"}:
            if in_quantity:
                quantity += c
            else:
                name += c
            idx += 1
            continue
        # ensure w/ '7 to 8 cups', the 'to 8` is added to the quantity portion
        # of the result
        elif in_quantity and c == "t" and value[idx - 1 : idx + 3] == " to ":
            letters = "to "
            quantity += letters
            idx += len(letters)
            continue
        elif c.isspace():
            if in_quantity:
                quantity += c
            else:
                name += c
            idx += 1
            continue
        else:
            if _starts_with(value[idx:], _units_ws) and in_quantity:
                in_quantity = False
                eat_count = 0
                for unit in _larger_to_smaller_units_ws:
                    if _starts_with(value[idx:], unit):
                        eat_count = len(unit)
                        break
                for _ in range(eat_count):
                    quantity += value[idx]
                    idx += 1
                # parse the parens that can occur after quantities, like:
                #   12 ounces (about 4 to 6 thighs)
                if value[idx] == "(":
                    while value[idx] != ")":
                        quantity += value[idx]
                        idx += 1
                    quantity += value[idx]
                    idx += 1
                    # parse: 1/2 cup (8 Tablespoons)/115 grams butter
                    if value[idx] == "/":
                        in_quantity = True
                for conjunction in ("plus", "+"):
                    if _starts_with(value[idx:], conjunction):
                        quantity += conjunction
                        idx += len(conjunction)
                        in_quantity = True
                continue
            if in_quantity:
                in_quantity = False
            name += c
            idx += 1
            continue

    return (quantity.strip(), name.strip())


_NON_INGREDIENT_NAMES = frozenset(
    {"bone-in", "skin-on", "fresh", "cooked", "raw", "frozen", "skinless", "boneless"}
)


def _parse_name_description(text: str) -> tuple[str, str]:
    """
    Some basic heuristics to partition a string into a name and description pair
    """
    prefix = ""
    temp = []
    is_all_suffix_now = False
    for word in text.split():
        if word.endswith(",") and not is_all_suffix_now:
            word_stripped = word.removesuffix(",")
            if word_stripped in _NON_INGREDIENT_NAMES:
                temp.append(word)
                continue
            if _is_unit(word_stripped):
                temp.append(word_stripped)
                continue
            temp.append(word_stripped)
            prefix = " ".join(temp)
            is_all_suffix_now = True
            temp = []
        else:
            temp.append(word)
    suffix = " ".join(temp)
    if not prefix:
        return (suffix, "")
    return (prefix, suffix)


# For example:
#   "Kosher salt" -> "kosher salt"
#   "Some Other Stuff" -> "Some Other Stuff"
def _normalize_name(text: str) -> str:
    text = text.strip()
    words = iter(text.split())
    first_word = next(words, None)
    if first_word is None:
        # no first word for some reason
        return text
    first_word_is_upper = first_word[0].isupper()
    following_words_are_lower = all(x[0].islower() for x in words)
    if first_word_is_upper and following_words_are_lower:
        return text.lower()
    return text


def parse_ingredient(text: str) -> _IngredientResult:
    quantity_name, description = _parse_name_description(text)
    quantity, name = _parse_quantity_name(quantity_name)
    is_optional = "optional" in text.lower()

    if is_optional:
        for suffix in (
            "(optional)",
            "optional",
            "[optional]",
            "Optional",
            "(Optional)",
            "[Optional]",
        ):
            if description.endswith(suffix):
                description = description[: -len(suffix)]
                break
            elif name.endswith(suffix):
                name = name[: -len(suffix)]
                break

    return _IngredientResult(
        # Some seems like a good default instead of empty string. Empty string
        # looks a little weird in the UI.
        quantity=quantity.strip() or "some",
        name=_normalize_name(name),
        description=description.strip(),
        optional=is_optional,
    )
