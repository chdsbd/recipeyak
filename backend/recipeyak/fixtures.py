"""
Create test data
"""
from recipeyak.models.ingredient import Ingredient
from recipeyak.models.note import Note
from recipeyak.models.recipe import Recipe
from recipeyak.models.section import Section
from recipeyak.models.step import Step
from recipeyak.models.team import Team
from recipeyak.models.user import User
from recipeyak.versioning import save_recipe_version


def create_user(*, email: str = "j.person@example.com") -> User:
    return User.objects.create_user(email=email)


def create_team(*, user: User) -> Team:
    team = Team.objects.create(name="Recipe Yak Team")
    team.force_join_admin(user=user)
    user.schedule_team = team
    user.save()
    return team


def create_recipe(*, team: Team, user: User) -> Recipe:
    recipe = Recipe.objects.create(
        name="Recipe name",
        author="Recipe author",
        source="www.exmple.com",
        time="1 hour",
        team=team,
    )

    Section.objects.create(recipe=recipe, title="a section title", position="a")
    Ingredient.objects.create(
        quantity="1 lbs",
        name="egg",
        description="scrambled",
        position="b",
        recipe=recipe,
    )

    Section.objects.create(recipe=recipe, title="a diff section", position="b")

    Ingredient.objects.create(
        quantity="2 tbs", name="soy sauce", position="c", recipe=recipe
    )

    Step.objects.create(
        text="Place egg in boiling water and cook for ten minutes",
        position="a",
        recipe=recipe,
    )

    Step.objects.create(
        text="serve",
        position="b",
        recipe=recipe,
    )

    Note.objects.create(recipe=recipe, text="Use a small pot.", created_by=user)
    Note.objects.create(
        recipe=recipe,
        text="I'd say we should use an even smaller pot next time.",
        created_by=user,
        last_modified_by=user,
    )

    # mimic what we do in the API
    save_recipe_version(recipe_id=recipe.id, actor=user)

    return recipe
