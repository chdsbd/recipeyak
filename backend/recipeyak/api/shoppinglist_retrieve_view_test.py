import json
from datetime import date, timedelta
from decimal import Decimal
from typing import Any

import pytest
from django.test.client import Client

from recipeyak.combine import Ingredient as IngredientCumin
from recipeyak.combine import (
    Quantity,
    combine_ingredients,
)
from recipeyak.fixtures import (
    create_ingredient,
    create_recipe,
    create_team,
    create_user,
)
from recipeyak.models import Ingredient, Recipe, ShoppingList, Team, User
from recipeyak.parsing import Unit

pytestmark = pytest.mark.django_db


def test_fetching_shoppinglist(
    client: Client, user: User, team: Team, recipe: Recipe
) -> None:
    assert ShoppingList.objects.count() == 0

    client.force_login(user)
    start = date(1976, 7, 6)
    end = start + timedelta(days=1)
    params = {"start": start, "end": end}

    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 200
    assert res.json()["ingredients"] == {}
    assert ShoppingList.objects.count() == 1

    recipe.schedule(user=user, on=start, team=team)

    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 200
    assert res.json()["ingredients"] == {
        "egg": {
            "category": "dairy",
            "quantities": [{"quantity": "1", "unit": "POUND", "unknown_unit": None}],
        },
        "soy sauce": {
            "category": "condiments",
            "quantities": [
                {"quantity": "2", "unit": "TABLESPOON", "unknown_unit": None}
            ],
        },
    }

    assert ShoppingList.objects.count() == 2
    shopping_list = ShoppingList.objects.order_by("-created").first()
    assert shopping_list is not None
    assert json.loads(shopping_list.ingredients) == res.json()["ingredients"]


def test_fetching_shoppinglist_with_team_recipe(
    client: Client, team: Team, user: User, recipe: Recipe
) -> None:
    client.force_login(user)

    assert team.is_member(user)

    start = date(1976, 7, 6)
    end = start + timedelta(days=1)
    params = {"start": start, "end": end}

    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 200
    assert res.json()["ingredients"] == {}

    recipe.schedule(user=user, on=start, team=team)

    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 200
    assert res.json()["ingredients"] != []

    assert res.json()["ingredients"] == {
        "egg": {
            "category": "dairy",
            "quantities": [{"quantity": "1", "unit": "POUND", "unknown_unit": None}],
        },
        "soy sauce": {
            "category": "condiments",
            "quantities": [
                {"quantity": "2", "unit": "TABLESPOON", "unknown_unit": None}
            ],
        },
    }


def test_fetching_shoppinglist_with_invalid_dates(
    user: User, team: Team, client: Client
) -> None:
    params = {"start": "", "end": "invalid date"}
    client.force_login(user)
    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 400


@pytest.mark.parametrize("quantity", ["sprinkle", "some"])
def test_scheduling_multiple_times_some_ingredient(
    quantity: str, user: User, team: Team, client: Client
) -> None:
    """
    with an ingredient of quantity sprinkle that we add to the cart multiple
    times shouldn't become sprinklesprinklesprinkle
    """

    name = "Recipe name"
    author = "Recipe author"
    recipe = Recipe.objects.create(name=name, author=author, team=team)

    Ingredient.objects.create(
        quantity=quantity, name="black pepper", position=10.0, recipe=recipe
    )

    start = date(1976, 7, 6)
    recipe.schedule(user=user, on=start, team=team)

    end = start + timedelta(days=1)
    params = {"start": start, "end": end}
    client.force_login(user)
    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 200
    assert res.json()["ingredients"] == {
        "black pepper": {
            "category": "spices",
            "quantities": [{"quantity": "1", "unit": "SOME", "unknown_unit": None}],
        }
    }


@pytest.mark.parametrize(
    "ingredients,expected",
    [
        (
            [("1/2", "lemon"), ("1", "lemon"), ("2", "lemons")],
            {"lemons": [Quantity(quantity=Decimal(3.5), unit=Unit.NONE)]},
        ),
        (
            [("1", "bay leaf"), ("4", "bay leaves")],
            {"bay leaves": [Quantity(quantity=Decimal(5), unit=Unit.NONE)]},
        ),
        (
            [("1", "large tomato"), ("2", "large tomatoes")],
            {"large tomatoes": [Quantity(quantity=Decimal(3), unit=Unit.NONE)]},
        ),
        (
            [("4-5", "medium button mushrooms"), ("4-5", "medium button mushrooms")],
            {
                "medium button mushrooms": [
                    Quantity(quantity=Decimal(10), unit=Unit.NONE)
                ]
            },
        ),
        (
            [
                ("1 tablespoon", "black pepper"),
                ("2 tablespoon", "black pepper"),
                ("sprinkle", "black pepper"),
                ("some", "black pepper"),
            ],
            {
                "black pepper": [
                    Quantity(quantity=Decimal(3), unit=Unit.TABLESPOON),
                    Quantity(quantity=Decimal(2), unit=Unit.SOME),
                ]
            },
        ),
        (
            [
                ("16", "basil leaves"),
                ("some", "basil leaves"),
                ("sprinkle", "basil leaves"),
            ],
            {
                "basil leaves": [
                    Quantity(quantity=Decimal(16), unit=Unit.NONE),
                    Quantity(quantity=Decimal(2), unit=Unit.SOME),
                ]
            },
        ),
        (
            [
                ("1 tablespoon", "extra-virgin olive oil"),
                ("8 tablespoons", "extra virgin olive oil"),
            ],
            {
                "extra virgin olive oil": [
                    Quantity(quantity=Decimal(9), unit=Unit.TABLESPOON)
                ]
            },
        ),
        (
            [("1", "garlic clove")],
            {"garlic clove": [Quantity(quantity=Decimal(1), unit=Unit.NONE)]},
        ),
        (
            [("8", "Garlic Cloves"), ("1", "garlic clove")],
            {"garlic cloves": [Quantity(quantity=Decimal(9), unit=Unit.NONE)]},
        ),
        (
            [("2 Tablespoons", "scallions"), ("4", "scallions")],
            {
                "scallions": [
                    Quantity(quantity=Decimal(2), unit=Unit.TABLESPOON),
                    Quantity(quantity=Decimal(4), unit=Unit.NONE),
                ]
            },
        ),
        (
            [("2 tbs", "soy sauce")],
            {"soy sauce": [Quantity(quantity=Decimal(2), unit=Unit.TABLESPOON)]},
        ),
        (
            [("2 lbs", "tomato"), ("1 kg", "tomato")],
            {
                "tomato": [
                    Quantity(
                        quantity=Decimal("4.204622864866847967783965457"),
                        unit=Unit.POUND,
                    )
                ]
            },
        ),
        (
            [("1 teaspoon + 1 Tablespoon", "ginger"), ("1 teaspoon", "Ginger")],
            {"ginger": [Quantity(quantity=Decimal(5), unit=Unit.TEASPOON)]},
        ),
        (
            [("1/2 cup", "scallions")],
            {"scallions": [Quantity(quantity=Decimal(0.5), unit=Unit.CUP)]},
        ),
        (
            [("Â½ cup", "scallions")],
            {"scallions": [Quantity(quantity=Decimal(0.5), unit=Unit.CUP)]},
        ),
        (
            [
                ("2", "lemons"),
                ("1", "lemon"),
                ("1 large", "lemon"),
                ("1/2", "lemon"),
                ("some", "lemon"),
            ],
            {
                "lemons": [
                    Quantity(quantity=Decimal(4.5), unit=Unit.NONE),
                    Quantity(quantity=Decimal(1), unit=Unit.SOME),
                ]
            },
        ),
    ],
)
def test_combine_ingredients(
    ingredients: list[tuple[str, str]], expected: dict[str, list[Quantity]]
) -> None:
    ingres = []
    position = 1
    for quantity, name in ingredients:
        ingres.append(IngredientCumin(quantity=quantity, name=name))
        position += 1

    assert combine_ingredients(ingres) == expected


def test_combining_feta(user: User, team: Team, client: Client) -> None:
    """
    ensure the singularize function doesn't result in feta becoming fetum along
    with some other troublesome examples
    """

    name = "empty recipe"
    empty_recipe = Recipe.objects.create(name=name, team=team)

    position = 11.0

    ingredients = [
        ("some", "feta"),
        ("1.25 cups", "all-purpose flour"),
        ("some", "katamata olives"),
        ("some", "red pepper flakes"),
        ("2 tablespoon", "molasses"),
        # duplicate the ingredient to replicate scheduling the recipe multiple
        # times
        ("1 bag", "tortilla chips"),
        ("1 bag", "tortilla chips"),
        ("1 container", "tortilla chips"),
    ]

    for quantity, name in ingredients:
        Ingredient.objects.create(
            quantity=quantity, name=name, position=position, recipe=empty_recipe
        )
        position += 10

    start = date(1976, 7, 6)
    empty_recipe.schedule(user=user, on=start, team=team)

    end = start + timedelta(days=1)
    params = {"start": start, "end": end}
    client.force_login(user)
    res = client.get("/api/v1/shoppinglist/", params)
    assert res.status_code == 200

    assert res.json()["ingredients"] == {
        "all purpose flour": {
            "category": "baking",
            "quantities": [{"quantity": "1.25", "unit": "CUP", "unknown_unit": None}],
        },
        "feta": {
            "category": "cheese",
            "quantities": [{"quantity": "1", "unit": "SOME", "unknown_unit": None}],
        },
        "katamata olives": {
            "category": "condiments",
            "quantities": [{"quantity": "1", "unit": "SOME", "unknown_unit": None}],
        },
        "molasses": {
            "category": "baking",
            "quantities": [
                {"quantity": "2", "unit": "TABLESPOON", "unknown_unit": None}
            ],
        },
        "red pepper flakes": {
            "category": "spices",
            "quantities": [{"quantity": "1", "unit": "SOME", "unknown_unit": None}],
        },
        "tortilla chips": {
            "category": "dry goods",
            "quantities": [
                {"quantity": "2", "unit": "UNKNOWN", "unknown_unit": "bag"},
                {"quantity": "1", "unit": "UNKNOWN", "unknown_unit": "container"},
            ],
        },
    }


def test_fetching_team_shopping_list(
    client: Client, user: User, team: Team, recipe: Recipe
) -> None:
    start = date(1976, 7, 6)
    end = start + timedelta(days=1)
    params = {"start": start, "end": end}

    url = "/api/v1/shoppinglist/"
    client.force_login(user)
    res = client.get(url, params)
    assert res.status_code == 200
    assert res.json()["ingredients"] == {}

    recipe.schedule(on=start, team=team, user=user)
    res = client.get(url, params)
    assert res.status_code == 200
    assert len(res.json()["ingredients"]) == (
        recipe.ingredient_set.count()
    ), "only return the schedule recipe ingredients"


def test_fetching_shopping_list_with_small_decimals() -> None:
    client = Client()
    user = create_user()
    team = create_team(user=user)
    recipe = create_recipe(team=team, user=user)
    create_ingredient(recipe=recipe, quantity="1/3 cup", name="flour", position="c")
    client.force_login(user)

    start = date(1976, 7, 6)
    end = start + timedelta(days=1)

    recipe.schedule(on=start, team=team, user=user)

    res = client.get("/api/v1/shoppinglist/", {"start": start, "end": end})
    assert res.status_code == 200
    rendered_quantity = res.json()["ingredients"]["flour"]["quantities"][0]["quantity"]
    assert rendered_quantity == "1/3"


def test_shoppinglist_rounding_edge_cases(snapshot: Any) -> None:
    client = Client()
    user = create_user()
    team = create_team(user=user)
    recipe = create_recipe(team=team, user=user)
    create_ingredient(
        recipe=recipe, quantity="1/3 cup", name="vegetable oil", position="d"
    )
    create_ingredient(
        recipe=recipe, quantity="4 Tablespoons", name="vegetable oil", position="f"
    )

    create_ingredient(recipe=recipe, quantity="4 teaspoons", name="sugar", position="g")
    create_ingredient(recipe=recipe, quantity="3/4 cup", name="sugar", position="h")

    client.force_login(user)

    start = date(1976, 7, 6)
    end = start + timedelta(days=1)

    recipe.schedule(on=start, team=team, user=user)

    res = client.get("/api/v1/shoppinglist/", {"start": start, "end": end})
    assert res.status_code == 200
    rendered_quantity = res.json()
    assert rendered_quantity == snapshot()
