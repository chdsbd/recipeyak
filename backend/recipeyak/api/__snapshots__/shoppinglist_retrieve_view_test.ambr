# name: test_shoppinglist_rounding_edge_cases
  dict({
    'ingredients': dict({
      'egg': dict({
        'category': 'dairy',
        'quantities': list([
          dict({
            'quantity': '1',
            'unit': 'POUND',
            'unknown_unit': None,
          }),
        ]),
      }),
      'soy sauce': dict({
        'category': 'condiments',
        'quantities': list([
          dict({
            'quantity': '2',
            'unit': 'TABLESPOON',
            'unknown_unit': None,
          }),
        ]),
      }),
      'sugar': dict({
        'category': 'baking',
        'quantities': list([
          dict({
            'quantity': '40',
            'unit': 'TEASPOON',
            'unknown_unit': None,
          }),
        ]),
      }),
      'vegetable oil': dict({
        'category': 'condiments',
        'quantities': list([
          dict({
            'quantity': '9.333',
            'unit': 'TABLESPOON',
            'unknown_unit': None,
          }),
        ]),
      }),
    }),
    'recipes': list([
      dict({
        'recipeId': 2876,
        'recipeName': 'Recipe name',
        'scheduledRecipeId': 802,
      }),
    ]),
  })
# ---
